store/store.go

```diff
@@ -770,3 +774,31 @@ type IntegrityCheckResult struct {
 	Data interface{}
 	Err  error
 }
+
+const mySQLDeadlockCode = uint16(1213)
+
+// WithDeadlockRetry retries a given f if it throws a deadlock error.
+// After exceeding a given threshold, it breaks and propagates the error upwards.
```

Comment:
A couple of things:

- Is the threshold given? I would expect the caller to decide how much to try before eventually failing instead of relying on a hard-coded value.
- Are there cases where it makes more sense to pass a time value (e.g. a deadline) and rely on that rather than on a number of retries?

Reply:
- IMO, I am not seeing much value in making the threshold as a parameter. Letting the user pass the parameter just adds new API surface without much benefit. Do you have any specific use-cases in mind? The problem is that there is no ideal number here that can be left to the user. Increasing from 3 to 5 just moves the goal post further.

If the query deadlocks even after retrying 3 times, there is something seriously broken somewhere that needs to be fixed IMO. 

Of course, there maybe cases of high load where deadlock-able queries get executed repeatedly. But in that case, even 5 or 10 might not be good enough. And it might be better just to stop the query altogether to prevent additional load on the DB and look towards fixing the deadlock some other way.

- IMO, I don't think so. MySQL would rollback one transaction while letting the other pass through. So that automatically lets the retried transaction to go forward.

Thoughts?

Reply:
As far as my first point I think it's fine to just update the documentation for that function since the threshold is not really passed as an argument but defined in there.

The second point was about removing the need to re-try a fixed amount of times but instead try as many times as possible within the given deadline. This might give more control to the caller but as you said, if the point is just trying to fix the edge case, the fixed retries should be enough.

Reply:
Ah I misunderstood your 1st point. Updated the godocs. PTAL.